source/game/system/CourseMap.cc

```diff
@@ -182,6 +182,434 @@ f32 CourseMap::startTmp3() const {
+                MapdataCheckPoint *prev =
+                        i < checkpoint->prevCount() ? checkpoint->prevPoint(i) : nullptr;
+                for (size_t j = 0; j < prev->nextCount(); j++) {
+                    MapdataCheckPoint *next = j < prev->nextCount() ? prev->nextPoint(j) : nullptr;
+                    if (next == checkpoint) {
+                        continue;
+                    }
+
+                    id = findRecursiveSector(pos, 1, false, *next, checkpointCompletion, params);
+                    if (id != -1) {
+                        break;
+                    }
+                }
+            }
+        }
+
+        // If that fails, search next
+        if (id == -1) {
+            for (size_t i = 0; i < checkpoint->nextCount(); i++) {
+                MapdataCheckPoint *checkpoint_ =
+                        i < checkpoint->nextCount() ? checkpoint->nextPoint(i) : nullptr;
+                id = findRecursiveSector(pos, 1, false, *checkpoint_, checkpointCompletion, params);
+                if (id != -1) {
+                    break;
+                }
+            }
+        }
+
+        // If that fails, search prev
+        if (id == -1) {
+            for (size_t i = 0; i < checkpoint->prevCount(); i++) {
+                MapdataCheckPoint *checkpoint_ =
+                        i < checkpoint->prevCount() ? checkpoint->prevPoint(i) : nullptr;
+                id = findRecursiveSector(pos, 1, true, *checkpoint_, checkpointCompletion, params);
+                if (id != -1) {
+                    break;
+                }
+            }
+        }
+
+    default:
+        break;
+    }
+
+    // GLOBAL SEARCH
+    // if local search fails, remove depth limit and search all "loaded" checkpoints
+
+    // Step 1: Search all next checkpoints until player or key checkpoint is found
+    if (id == -1) {
+        for (size_t i = 0; i < checkpoint->nextCount(); i++) {
+            MapdataCheckPoint *checkpoint_ =
+                    i < checkpoint->nextCount() ? checkpoint->nextPoint(i) : nullptr;
+            id = findRecursiveSector(pos, -1, false, *checkpoint_, checkpointCompletion, 0);
+            if (id != -1) {
+                break;
+            }
+        }
+        // Step 2: Search all previous checkpoints until player or key checkpoint is found
+        if (id == -1) {
+            for (size_t i = 0; i < checkpoint->prevCount(); i++) {
+                MapdataCheckPoint *checkpoint_ =
+                        i < checkpoint->prevCount() ? checkpoint->prevPoint(i) : nullptr;
+                id = findRecursiveSector(pos, -1, true, *checkpoint_, checkpointCompletion, 0);
+                if (id != -1) {
+                    break;
+                }
+            }
+        }
+    }
+
+    // ONLINE PLAYERS ONLY
+    // if both local and global searches fail, just iterate through ALL checkpoints (ignoring kcps)
+    // until player is found or last checkpoint reached
+    if (isRemote && id == -1) {
```

Comment:
Add early return to remove some indentation here
```suggestion
    if (!isRemote || id != -1) {
        return id;
    }
```